zookeeper-server/src/main/java/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java

```diff
@@ -499,23 +499,28 @@ public void save(
      * @return true if able to truncate the log, false if not
      * @throws IOException
      */
-    public boolean truncateLog(long zxid) throws IOException {
-        // close the existing txnLog and snapLog
-        close();
-
-        // truncate it
-        FileTxnLog truncLog = new FileTxnLog(dataDir);
-        boolean truncated = truncLog.truncate(zxid);
-        truncLog.close();
-
-        // re-open the txnLog and snapLog
-        // I'd rather just close/reopen this object itself, however that
-        // would have a big impact outside ZKDatabase as there are other
-        // objects holding a reference to this object.
-        txnLog = new FileTxnLog(dataDir);
-        snapLog = new FileSnap(snapDir);
-
-        return truncated;
+    public boolean truncateLog(long zxid) {
+        try {
+            // close the existing txnLog and snapLog
+            close();
+
+            // truncate it
+            FileTxnLog truncLog = new FileTxnLog(dataDir);
+            boolean truncated = truncLog.truncate(zxid);
+            truncLog.close();
+
+            // re-open the txnLog and snapLog
+            // I'd rather just close/reopen this object itself, however that
+            // would have a big impact outside ZKDatabase as there are other
+            // objects holding a reference to this object.
+            txnLog = new FileTxnLog(dataDir);
+            snapLog = new FileSnap(snapDir);
+
+            return truncated;
+        } catch (IOException e) {
```

Comment:
I'm wondering if IOException is sufficient. If you look at 
https://github.com/apache/zookeeper/blob/c234848caef0cc920f19befd1b0b14251b17da92/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Follower.java#L128
It's catching an Exception, so basically any exception other than IOException occurring here, could kick us back into the same situation. In fact, even the caller of Follower.followLeader() catches Exception. They're all over QuorumPeer, with "Unexpected exception" log warnings.

I think as part of the same fix, these "Unexpected exception" errors should also take down the process. wdyt?

Reply:
I'll check it, but currently I don't see any reason why not change it to `Exception`.

Reply:
I think it would be better to change the behaviour in QuorumPeer to requestSystemExit any time it gets an unexpected exception. QuorumPeer runs in a loop starting Learners and Leaders etc. There are objects shared between iterations of these loops, and as far I can see little checking that the objects are in a sane state if an exception occurred. Of course, I this is a bigger change, so perhaps it can be done later.

Reply:
I also think this should be another ticket/improvement to fix